lua/packer.lua

```diff
@@ -181,6 +210,8 @@ manage = function(plugin)
   plugin.name = path
   plugin.path = path
 
+  plugin.commit = plugins_snapshot[plugin.short_name]
```

Comment:
I might be missing something, but: what if there's no commit for a plugin in a snapshot, either if the user has no snapshot or if there is a new plugin in the specification not included in an old snapshot?

Reply:
I'm also a bit curious how this will interact with currently specified plugins/config values. It seems as though they can override the snapshotted values, but is that what we want? We should discuss the desired semantics of snapshots. I'm not familiar with this feature in e.g. `plug`, so this may already be a solved problem.

Reply:
> I might be missing something, but: what if there's no commit for a plugin in a snapshot, either if the user has no snapshot or if there is a new plugin in the specification not included in an old snapshot?

If a user doesn't have a plugin's snapshot it will be `plugin.commit = nil` (which is already the default behaviour when you don't specify the commit in `use`), so it wouldn't make any difference imo

Reply:
> I'm also a bit curious how this will interact with currently specified plugins/config values. It seems as though they can override the snapshotted values, but is that what we want? We should discuss the desired semantics of snapshots. I'm not familiar with this feature in e.g. `plug`, so this may already be a solved problem.

Do you mean if the `plugin.commit` has already a value, should it be rolled back to the snapshotted state?
Didn't really think about that so I guess you have a point.
Maybe we should just ignore the plugins which have commit `specified`? It would make more sense to me since if I want to specify a specific commit version of a package I would like to use that in any case.

Reply:
> I'm also a bit curious how this will interact with currently specified plugins/config values. It seems as though they can override the snapshotted values, but is that what we want? We should discuss the desired semantics of snapshots. I'm not familiar with this feature in e.g. `plug`, so this may already be a solved problem.

If I remember correctly when you take a snapshot using `vim-plug` it actually generates a vimscript file which you can `:source` and you have rolled back to the previous state.

In the beginning I wanted to make something similar, but I didn't like that you had to specify the snapshot's location and that you had to source it manually.

But it could be interesting to have a PackerRollback <snapshot> to load a snapshot only in the current session.

Reply:
Although it's not really intended for this purpose, `packer` relies on `vim.inspect` to serialize tables for e.g. generating the compiled file. That said, it seems like all we'd really need is to output the plugin names and commits and write a `rollback` function to checkout those commits in the relevant directories. I think those semantics might be cleaner than figuring out the potential interactions between a snapshot and a "fresh" config, tbh.